docs/DeveloperGuide.adoc

```diff
@@ -239,7 +239,312 @@ Classes used by multiple components are in the `seedu.addressbook.commons` packa
+'''
+The following sequence diagrams shows you how `find` and `undosearch` commands utilize `SearchHistoryManager` to perform Search Pruning.
+
+ - `find` command +
+
+image::SearchPruningSequenceDiagram.png[width="800"]
+
+ - `undosearch` command +
+
+image::undoSearchSequenceDiagram.png[width="800"]
+
+
+==== Design Considerations
+
+Aspect: What data is stored in search history stack +
+
+* **Alternative 1(current choice):**  Save a Stack of Predicates
+** Pros:
+    . Does not need to store the data in search history explicitly which saves memory +
+    . Any form of Search Pruning done with Predicates can reuse `SearchHistoryManager` class +
+
+** Cons:
+. Need to understand how `Predicate` works before utilizing this Class.
+. `Predicate` objects by itself does not perform the Search Pruning. In NUSSU Connect, we have to pass the `Predicate`
+   into a `FilteredList<T>` object to do the Search Pruning.
+
+* ** Alternative 2: ** Save a Stack of Lists containing Person objects in search history
+** Pros:
+    . It is easy to understand that we are filtering according to Person objects from `SearchHistoryManager` class
+
+** Cons:
+    . More memory is required as Person objects has to be duplicated multiple times into a new List.
+    . Class is inextensible to store other objects other than those from a Person classes.
+
+
+=== Add Skill Level Command
+
+==== Current Implementation
+
+The add skill mechanism builds on the `addressBookParser`. This as well
+as it's subclass `addSkillCommandParser` ensures that the correct number of arguments
+is given to the command.
+
+The following shows how the application handles the request to change a skill in one particular scenario.
+
+Step 1. The user launches the application. The application boots up and lists all members.
+
+Step 2. The user locates the person he wants to add on at Index 4. They execute the `asl 4 s/Photography l/60` command.
+
+Step 3.'LogicManager' calls the 'parseCommand' in the 'AddressBookParser', which calls `AddSkillCommandParser` to
+parse it.
+
+Step 4. After parsing, the command is sent to the `Model` which alters the `Person` object by modifying their
+`Skill` and `SkillLevel` properties.
+
+Step 5. The result is encapsulated as a `CommandResult` object which is passed back to the `UI`.
+
+
+Before executing the command:
+
+image::aslbefore.png[width="800"]
+After executing the command:
+
+image::aslafter.png[width="800"]
+
+==== Alternate implementations
+
+We considered two different ways to implement the Skill Class.
+
+* **Alternative 1**: Combining both Skill and SkillLevel properties together into a single class.
+
+** Pros: Resembles the real world, as there is a one-to-one mapping of Skill to SkillLevel.
+** Cons: Harder to test, and violates Single Responsiblity Principle.
+
+* **Alternative 2 (Current Choice)**: Separating the Skill and SkillLevel classes into different classes.
+
+** Pros: Easier to test.
+** Cons: Adds to the number of classes unnecessarily.
```

Comment:
Note that `SkillLevel` would be an association class. The additional information can be encapsulated in the class.